### PR TITLE
Synchronous pipelines

### LINKED ISSUE #166: Alarms are acknowledged even when no message has been successfully sent
Presumably happens only when the message sending itself fails

### LINKED ISSUE #170: Pipeline drift doesn't quite drift properly
We should add some extra drift for cases where we don't get new values from the database, to help realign the readout and pipeline phases.

### LINKED ISSUE #177: Pipelines do not stop sleeping
After an alarm a pipeline remains silent for ever

### LINKED ISSUE #178: Alarm for not responding should have same recurrance before sending alarm as for values out of range
Otherwise LM alarms are essentially useless

### PR DESCRIPTION
**Pull request by darrylmasson**
Adds support for a new SourceNode that receives data directly from Sensors, rather than having to go through the database as an intermediate step. Envisioned for Control and Convert pipelines. Sensors send their values both to Influx and also to the Control and Convert pipeline monitors. This inter-communication is not done through the hypervisor to reduce bottlenecking (note - do some benchmarking to see how high the rate can go).

Also adds a variety of other things related to hypervising, and a few optimizations to reduce the number of database queries.

Also fixes #178, fixes #177, fixes #170, fixes #166

### CODE REVIEW

**Review by jarongrigat — APPROVED** [review]

File: Doberman/PipelineMonitor.py [hunk 1]
```diff
@@ -26,27 +28,56 @@ def start_pipeline(self, name):
...
+            if command.startswith('sensor_value'):
+                _, name, ts, value = command.split()
+                ts = float(ts)
+                if value == 'None':
+                    value = None
+                elif '.' in value:
+                    value = float(value)
+                else:
+                    value = int(value)
+                for listeners in self.listeners.get(name, []):
```
[the remark this answers is not in the document]
Comment (darrylmasson) [reply Doberman/PipelineMonitor.py:78]: Yes

File: Doberman/hypervisor.py [hunk 2]
```diff
@@ -98,18 +103,37 @@ def hypervise(self) -> None:
...
+        """
+        try:
+            hn, p = self.get_listener_address(thing)
+            with socket.create_connection((hn, p), timeout=0.1) as sock:
+                sock.sendall(b'ping')
+                if sock.recv(1024) != b'pong':
+                    return True
+        except Exception as e:
+            self.logger.debug(f'Caught a {type(e)} while pinging {thing}: {e}')
+            return True
```
[the remark this answers is not in the document]
Comment (darrylmasson) [reply Doberman/hypervisor.py:128]: `True` means we should restart `thing`. The only things that get pinged are either the three Pipeline Monitors, which are always supposed to be running, or managed DeviceMonitors, which are supposed to be running. There are two likely exceptions here, one being a `ValueError` if there is no listener info for `thing`, in which case it means it needs to get restarted, or we get a `ConnectionRefusedError` which means the thing might be running but it isn't listening, in which case it needs to get restarted.

[the remark this answers is not in the document]
Comment (darrylmasson) [reply Doberman/hypervisor.py:128]: It's possible that after testing this needs to get changed.

[the remark this answers is not in the document]
Comment (jarongrigat) [reply Doberman/hypervisor.py:128]: Oh yes, I misread something, apparently.

**Pull request merged**